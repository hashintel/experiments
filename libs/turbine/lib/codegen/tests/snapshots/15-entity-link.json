[
  {
    "$id": "http://localhost:3000/@alice/types/entity-type/partOf/v/1",
    "$schema": "https://blockprotocol.org/types/modules/graph/0.3/schema/entity-type",
    "allOf": [
      {
        "$ref": "https://blockprotocol.org/@blockprotocol/types/entity-type/link/v/1"
      }
    ],
    "description": null,
    "kind": "entityType",
    "links": {},
    "properties": {},
    "required": [],
    "title": "Part Of",
    "type": "object"
  }
]
