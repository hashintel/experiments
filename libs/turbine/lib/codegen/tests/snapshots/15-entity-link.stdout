localhost_3000/alice/entity/part_of.rs

use alloc::string::String;
use core::marker::PhantomData;

use error_stack::{Report, Result, ResultExt as _};
use hashbrown::HashMap;
use serde::Serialize;
use turbine::{
    entity::{Entity, LinkData},
    url, BaseUrl, EntityLink, EntityProperties, EntityType, EntityTypeMut, EntityTypeRef,
    GenericEntityError, OptionalEntityLink, PropertyType as _, PropertyTypeMut as _,
    PropertyTypeRef as _, Type, TypeMut, TypeRef, TypeUrl, VersionedUrlRef,
};
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Properties;
turbine::serialize_compat!(Properties);
impl Properties {
    fn try_from_value(
        mut properties: HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        Ok(Self)
    }

    fn as_mut(&mut self) -> PropertiesMut<'_> {
        PropertiesMut(PhantomData)
    }

    fn as_ref(&self) -> PropertiesRef<'_> {
        PropertiesRef(PhantomData)
    }
}
///Part Of
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PartOf {
    pub properties: Properties,
    pub link_data: LinkData,
}
pub type PartOfV1 = PartOf;
impl TypeUrl for PartOf {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/partOf/" / v / 1u32);
}
impl Type for PartOf {
    type Mut < 'a > = PartOfMut < 'a > where Self : 'a ;
    type Ref < 'a > = PartOfRef < 'a > where Self : 'a ;

    fn as_mut(&mut self) -> Self::Mut<'_> {
        PartOfMut {
            properties: self.properties.as_mut(),
            link_data: &mut self.link_data,
        }
    }

    fn as_ref(&self) -> Self::Ref<'_> {
        PartOfRef {
            properties: self.properties.as_ref(),
            link_data: &self.link_data,
        }
    }
}
impl EntityType for PartOf {
    type Error = GenericEntityError;

    fn try_from_entity(value: Entity) -> Option<Result<Self, Self::Error>> {
        if Self::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = Properties::try_from_value(value.properties.0);
        let link_data = value
            .link_data
            .ok_or_else(|| Report::new(GenericEntityError::ExpectedLinkData));
        match turbine::fold_tuple_reports((properties, link_data)) {
            Err(error) => Some(Err(error)),
            Ok((properties, link_data)) => Some(Ok(Self {
                properties,
                link_data,
            })),
        }
    }
}
impl EntityProperties for PartOf {
    type Properties = Properties;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl EntityLink for PartOf {
    fn link_data(&self) -> &LinkData {
        &self.link_data
    }
}
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PropertiesRef<'a>(pub PhantomData<&'a ()>);
turbine::serialize_compat!(PropertiesRef<'a>);
impl<'a> PropertiesRef<'a> {
    fn try_from_value(
        properties: &'a HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        Ok(Self(PhantomData))
    }

    fn into_owned(self) -> Properties {
        Properties
    }
}
///Part Of
#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PartOfRef<'a> {
    pub properties: PropertiesRef<'a>,
    pub link_data: &'a LinkData,
}
pub type PartOfV1Ref<'a> = PartOfRef<'a>;
impl TypeUrl for PartOfRef<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/partOf/" / v / 1u32);
}
impl TypeRef for PartOfRef<'_> {
    type Owned = PartOf;

    fn into_owned(self) -> Self::Owned {
        PartOf {
            properties: self.properties.into_owned(),
            link_data: self.link_data.clone(),
        }
    }
}
impl<'a> EntityTypeRef<'a> for PartOfRef<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesRef::try_from_value(&value.properties.0);
        let link_data = value
            .link_data
            .as_ref()
            .ok_or_else(|| Report::new(GenericEntityError::ExpectedLinkData));
        match turbine::fold_tuple_reports((properties, link_data)) {
            Err(error) => Some(Err(error)),
            Ok((properties, link_data)) => Some(Ok(Self {
                properties,
                link_data,
            })),
        }
    }
}
impl<'a> EntityProperties for PartOfRef<'a> {
    type Properties = PropertiesRef<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl EntityLink for PartOfRef<'_> {
    fn link_data(&self) -> &LinkData {
        self.link_data
    }
}
#[derive(Debug)]
pub struct PropertiesMut<'a>(pub PhantomData<&'a ()>);
turbine::serialize_compat!(PropertiesMut<'a>);
impl<'a> PropertiesMut<'a> {
    fn try_from_value(
        properties: &'a mut HashMap<String, serde_json::Value>,
    ) -> Result<Self, GenericEntityError> {
        Ok(Self(PhantomData))
    }

    fn into_owned(self) -> Properties {
        Properties
    }
}
///Part Of
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PartOfMut<'a> {
    pub properties: PropertiesMut<'a>,
    pub link_data: &'a mut LinkData,
}
pub type PartOfV1Mut<'a> = PartOfMut<'a>;
impl TypeUrl for PartOfMut<'_> {
    type InheritsFrom = ();

    const ID: VersionedUrlRef<'static> =
        url!("http://localhost:3000/@alice/types/entity-type/partOf/" / v / 1u32);
}
impl TypeMut for PartOfMut<'_> {
    type Owned = PartOf;

    fn into_owned(self) -> Self::Owned {
        PartOf {
            properties: self.properties.into_owned(),
            link_data: self.link_data.clone(),
        }
    }
}
impl<'a> EntityTypeMut<'a> for PartOfMut<'a> {
    type Error = GenericEntityError;

    fn try_from_entity(value: &'a mut Entity) -> Option<Result<Self, Self::Error>> {
        if Self::Owned::ID != value.metadata.entity_type_id {
            return None;
        }
        let properties = PropertiesMut::try_from_value(&mut value.properties.0);
        let link_data = value
            .link_data
            .as_mut()
            .ok_or_else(|| Report::new(GenericEntityError::ExpectedLinkData));
        match turbine::fold_tuple_reports((properties, link_data)) {
            Err(error) => Some(Err(error)),
            Ok((properties, link_data)) => Some(Ok(Self {
                properties,
                link_data,
            })),
        }
    }
}
impl<'a> EntityProperties for PartOfMut<'a> {
    type Properties = PropertiesMut<'a>;

    fn properties(&self) -> &Self::Properties {
        &self.properties
    }
}
impl EntityLink for PartOfMut<'_> {
    fn link_data(&self) -> &LinkData {
        &*self.link_data
    }
}
